lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -1150,6 +1191,180 @@ void SESERegionBuilder::ensureSingleExitFromLoops() {
+  // value before the loop gets executed. However, note that on all paths from
+  // `header` to `newLatch`, the second argument is always defined. Therefore,
+  // we can eliminate the `undef` by simply unrolling the loop body from
+  // `header` to `newLatch` once. We do not need to  clone `newHeader`.
+  //
+  // Given that we do not clone `newHeader`, references to the arguments `i1`
+  // and `i5` of `newHeader` will not be cloned as well. For example, in the
+  // unrolled body of the loop, `break` will be cloned as follows (prime refers
+  // to the clones):
+  //    break':
+  //      br newLatch'(i1, i2', false)
+  //
+  // Note that `i1` still refers to the original argument of `newHeader`.
+  // `newHeader` occurs after the unrolled loop body (in CFG) and hence, won't
+  // dominate `break'` in the unrolled loop body. We fix this by
+  // picking a value that is equivalent to the corresponding argument and
+  // dominates `break'`. (Note that it is enough to search for equivalent values
+  // among the immediate predecessors of the `newLatch'` block.)
+  //
+  // (1) Unroll the loop body once.
+  //
+  BasicBlockCloner cloner(*currentFn);
+  // Setup cloner so that newHeader's arguments are replaced with values in
+  // preheader.
+  SILBasicBlock *newHeader = loop->getHeader();
+  auto preheaderTermInst = dyn_cast<BranchInst>(preheader->getTerminator());
+  assert(preheaderTermInst && "Preheader of a loop has a non-branch terminator");
+  for (unsigned argIndex = 0; argIndex < oldHeaderNumArgs; ++argIndex) {
+    auto preheaderArg = preheaderTermInst->getArg(argIndex);
+    auto newHeaderArg = newHeader->getArgument(argIndex);
+    cloner.updateValueMap(newHeaderArg, preheaderArg);
+  }
+  // Clone everything except the new header. We should traverse the
+  // blocks in depth first order to ensure values are cloned before they are used.
+  SmallPtrSet<SILBasicBlock *, 32> worklist;
+  SmallVector<SILBasicBlock *, 32> initializedBlocks;
+  worklist.insert(header);
+  while (!worklist.empty()) {
+    SILBasicBlock *current = *worklist.begin();
+    worklist.erase(current);
+    cloner.initBlock(current);
+    initializedBlocks.push_back(current);
+    for (SILBasicBlock *succ : current->getSuccessorBlocks()) {
+      // Skip if succ is not a part of the loop, is already cloned, or
+      // is the new preheader.
+      if (!loop->contains(succ) || cloner.remapBasicBlock(succ) != succ ||
+          succ == newHeader) {
+        continue;
+      }
+      worklist.insert(succ);
+    }
+  }
+
+  SILLoop *parentLoop = loop->getParentLoop();
+  for (SILBasicBlock *bb : initializedBlocks) {
+    cloner.cloneBlock(bb);
+    SILBasicBlock *clonedBlock = cloner.cloneBlock(bb);
+    if (parentLoop) {
+      parentLoop->addBasicBlockToLoop(clonedBlock, LI->getBase());
+    }
+  }
+
+  // Get the clone for the original and new header.
+  SILBasicBlock *clonedHeader = cloner.remapBasicBlock(header);
```

Comment:
given "newHeader", it'd be less confusing if we change this to "oldHeader", and same for clonedHeader.

Reply:
Changed to clonedOldHeader. `header` is a class field.